text/0000-lifecycle-pkg-refactor.md

```diff
@@ -0,0 +1,174 @@
+├── ...
+```
+
+Here is the proposed package structure:
+
+```
+├── cmd
+│    ├── launcher
+│    ├── lifecycle
+├── internal
+│    ├── buildpack              # services for handling logic that branches on buildpack api - e.g., bom validator
+│    ├── encoding
+│    ├── image
+│    ├── io
+│    ├── platform               # services for handling logic that branches on platform api - e.g., metadata restorer, sbom restorer, potentially a new sbom copier, etc.
+│    └── str
+├── pkg
+│    ├── api                    # old api package with CNB logic that indicates supported and deprecated apis moved elsewhere
+│    ├── archive
+│    ├── buildpack
+│    │   ├── env
+│    │   ├── launch
+│    ├── lifecycle
+│    │   ├── layers
+│    │   ├── analyzer.go        # lifecycle logic moved to nested package
+│    │   ├── detector.go
+│    │   ├── rebaser.go
+│    │   ├── ...
+│    └── platform
+│        ├── auth
+│        ├── cache
+│        ├── inputs             # handles platform <-> lifecycle interface (args, flags)
+│        ├── launch
+│        │   └── inputs         # handles platform <-> launcher interface (args, flags)
+│        └── priv
+```
+
+## Considerations
+
+* The structure of the lifecycle should mirror the spec as much as possible. However we should recognize that the spec is not perfect. Also, while there is a `lifecycle` package, there is no "lifecycle spec" - the responsibilities of the lifecycle are mixed between the platform and buildpack specs (see below).
+* In general the `lifecycle` package should be more of an orchestrator - see [example](https://github.com/buildpacks/lifecycle/blob/b6803be364429689a818a5d0a09db42bd21b9995/analyzer.go#L52-L54). We should push as much logic as possible into services.
+* CNB business logic should live inside the `buildpack`, `lifecycle`, and `platform` packages (and their nested packages) or their `internal` equivalents. All other packages that are direct children of `pkg` or `internal` should be CNB-free.
+* Logic that branches on api version should be handled by an internal service nested under the appropriate package - e.g., logic that branches on buildpack api should live in `internal/buildpack` - see [example](https://github.com/buildpacks/lifecycle/blob/b6803be364429689a818a5d0a09db42bd21b9995/buildpack/bom.go#L14-L23).
+* Tests should serve as documentation for apis by enumerating which services are in use - see [example](https://github.com/buildpacks/lifecycle/blob/a457244e2785afd342fe511563445a4fa4554e3e/platform/inputs/analyzer_factory_test.go#L353-L375).
+* To keep the launcher as lean as possible, we should minimize the packages it depends on (hence `platform/launch/inputs`).
+
+## Aligning with the spec
+
+The spec has this structure (abridged):
+
+Platform:
+
+  - Stacks
+  - Lifecycle
+    - Operations
+      - Build
+      - Rebase
+      - Launch
+    - Platform inputs and outputs (args, flags, env vars, mounts, exit codes)
+    - Registry auth
+  - Buildpacks directory layout
```

Comment:
It has always sort of bothered me that the `buildpack` package contains [`DirBuildpackStore`](https://github.com/buildpacks/lifecycle/blob/b6803be364429689a818a5d0a09db42bd21b9995/buildpack/store.go#L29-L38) and I think this is why!